src/main/scala/nl.knaw.dans.easy.solr4files/package.scala

```diff
@@ -15,20 +15,144 @@
  */
 package nl.knaw.dans.easy
 
-import java.nio.file.Paths
+import java.io.File
+import java.net.{ URL, URLDecoder }
 
-import scala.util.{Failure, Success, Try}
+import nl.knaw.dans.lib.error._
+import nl.knaw.dans.lib.logging.DebugEnhancedLogging
+import org.apache.commons.io.FileUtils.readFileToString
+import org.apache.solr.common.util.NamedList
 
-package object solr4files {
+import scala.annotation.tailrec
+import scala.util.{ Failure, Success, Try }
+import scala.xml.{ Elem, Node, XML }
+import scalaj.http.{ Http, HttpResponse }
 
-  implicit class TryExtensions2[T](val t: Try[T]) extends AnyVal {
-    // TODO candidate for dans-scala-lib
-    def unsafeGetOrThrow: T = {
-      t match {
-        case Success(value) => value
-        case Failure(throwable) => throw throwable
+package object solr4files extends DebugEnhancedLogging {
+
+  type FeedBackMessage = String
+  type SolrLiterals = Seq[(String, String)]
+  type FileToShaMap = Map[String, String]
+  type VocabularyMap = Map[String, String]
+
+  case class HttpStatusException(msg: String, response: HttpResponse[String])
+    extends Exception(s"$msg - ${ response.statusLine }, details: ${ response.body }")
+
+  case class SolrStatusException(namedList: NamedList[AnyRef])
+    extends Exception(s"solr update returned: ${ namedList.asShallowMap().values().toArray().mkString }")
+
+  case class SolrDeleteException(bagId: String, cause: Throwable)
+    extends Exception(s"solr delete of bag $bagId failed with ${ cause.getMessage }", cause)
+
+  case class SolrUpdateException(solrId: String, cause: Throwable)
+    extends Exception(s"solr update of file $solrId failed with ${ cause.getMessage }", cause)
+
+  case class SolrCommitException(cause: Throwable)
+    extends Exception(cause.getMessage, cause)
+
+  case class MixedResultsException[T](results: Seq[T], thrown: Throwable)
+  // TODO evolve into candidate for dans.lib.error with takeUntilFailure
+    extends Exception(thrown.getMessage, thrown)
+
+  implicit class RichSeq[T](val left: Seq[T]) extends AnyVal {
+    def stats: String = {
+      val xs = left.groupBy(_.getClass.getSimpleName)
+      xs.keySet.map(className => s"${ xs(className).size } times $className").mkString(", ")
+    }
+  }
+
+  implicit class RichTryStream[T](val left: Seq[Try[T]]) extends AnyVal {
+
+    /** Typical usage: val (thrown, results) = ...toStream.map(TrySomething).takeUntilFailure */
+    def takeUntilFailure: Try[Seq[T]] = {
+      val it = left.iterator
+
+      @tailrec
+      def inner(result: List[T] = List.empty): Try[Seq[T]] = {
+        if (!it.hasNext) Success(result.reverse)
+        else it.next() match {
+          case Success(y) => inner(y :: result)
+          case Failure(t) =>
+            val seq: Seq[T] = result.reverse
```

Comment:
inline this line if possible (maybe it's actually needed because of the explicit type declaration???)